ts/src/program/index.ts

```diff
@@ -304,28 +319,42 @@ export class Program {
    * @param callback  The function to invoke whenever the event is emitted from
    *                  program logs.
    */
-  public addEventListener(
+   public addEventListener(
     eventName: string,
     callback: (event: any, slot: number) => void
-  ): number {
-    const eventParser = new EventParser(this._coder, this._programId);
-    return this._provider.connection.onLogs(this._programId, (logs, ctx) => {
-      if (logs.err) {
-        console.error(logs);
-        return;
-      }
-      eventParser.parseLogs(logs.logs, (event) => {
-        if (event.name === eventName) {
-          callback(event.data, ctx.slot);
+  ): void {
+    if (eventName in this._eventCallbacks){
+      throw new Error(`Event listener for $(eventName) already exists!`);
+    }
+
+    this._eventCallbacks[eventName] = callback;
+
+    if (this._onLogsSubscriptionId == undefined) {
```

Comment:
I kind of dislike having this logic inside addEventListener, and feel like it should belong in its own `addOnLogsSubscription` function, but this abstracts that logic away from the client and could be cleaner?